libvast/src/system/evaluator.cpp

```diff
@@ -0,0 +1,173 @@
+    auto result = caf::visit(*this, n.expr());
+    result.flip();
+    return result;
+  }
+
+  ids operator()(const predicate& pred) const {
+    auto i = xs.find(pred);
+    return i != xs.end() ? i->second.second : ids{};
+  }
+
+  const evaluator_state::sub_hits_map& xs;
+};
+
+} // namespace
+
+evaluator_state::evaluator_state(caf::event_based_actor* self) : self(self) {
+  // nop
+}
+
+void evaluator_state::init(caf::actor client, expression expr) {
+  this->client = std::move(client);
+  this->expr = std::move(expr);
+}
+
+void evaluator_state::handle_indexer_result(const predicate& pred,
+                                            const ids& result) {
+  VAST_DEBUG(self, "got new hits", result, " for predicate", pred);
+  if (auto ptr = sub_hits_of(pred); ptr == nullptr) {
+    VAST_ERROR(self, "got a result for an unknown predicate:", pred);
+  } else {
+    auto& [missing, accumulated_hits] = *ptr;
+    accumulated_hits |= result;
+    if (--missing == 0) {
+      VAST_DEBUG(self, "collected all INDEXER results for predicate", pred);
+      update();
+    }
+  }
+  decrement_pending();
+}
+
+void evaluator_state::handle_missing_indexer_result(const predicate& pred,
+                                                    const caf::error& err) {
+  VAST_IGNORE_UNUSED(err);
+  VAST_DEBUG(self, "got no result for predicate", pred, "INDEXER returned",
+             self->system().render(err));
+  if (auto ptr = sub_hits_of(pred); ptr == nullptr) {
+    VAST_ERROR(self, "got an error for an unknown predicate:", pred);
+  } else if (--ptr->first == 0) {
+    VAST_DEBUG(self, "collected all INDEXER results for predicate", pred);
+    update();
+  }
+  decrement_pending();
+}
+
+void evaluator_state::update() {
+  VAST_DEBUG(self, "got sub_hits:", sub_hits,
+             "expr_hits:", caf::visit(ids_evaluator{sub_hits}, expr));
+  auto delta = caf::visit(ids_evaluator{sub_hits}, expr) - hits;
+  if (any<1>(delta)) {
+    hits |= delta;
+    self->send(client, std::move(delta));
+  }
+}
+
+void evaluator_state::decrement_pending() {
+  // We're done evaluating if all INDEXER actors have reported their hits.
+  if (--pending_responses == 0) {
+    VAST_DEBUG(self, "completed expression evaluation");
+    self->send(client, done_atom::value);
+  }
+}
+
+evaluator_state::sub_hits_map::mapped_type*
+evaluator_state::sub_hits_of(const predicate& pred) {
+  auto i = sub_hits.find(pred);
+  return i != sub_hits.end() ? &i->second : nullptr;
+}
+
+caf::behavior evaluator(caf::stateful_actor<evaluator_state>* self,
+                        std::vector<caf::actor> indexers) {
+  // Skip incoming queries when already processing one.
+  return {[=](const expression& expr, caf::actor client) {
+    self->state.init(client, expr);
+    // TODO: we might want to locate the smallest subset of indexers (checking
+    //       whether the predicate could match the type of the indexer) instead
+    //       of always querying all indexers.
```

Comment:
This comment is important, but I think it's now at the wrong location because here we're dealing with type-erased actors. Perhaps move it into the parent context?

Reply:
There's no parent context yet. This PR only adds the EVALUATOR. I've removed the comment for now.